torch/nested/_internal/union_find.py

```diff
@@ -0,0 +1,172 @@
+import weakref
+
+import torch
+from torch.utils.weak import WeakTensorKeyDictionary
+from typing import *  # noqa: F403
+
+
+# Union find over versions of tensors
+# -----------------------------------
+# TensorUnionFind implements union-find over "versions of tensors". What this
+# means is that if TensorUnionFind holds a set of two tensors {`a`, `b`} with
+# `a` as the canonical entry, we should really think of it as holding {(`a`, 0),
+# (`b`, 0)} with (`a`, 0) as the canonical entry, where the 0s are the versions
+# of the tensors.
+#
+# If the tensor `a` is mutated, incrementing its version e.g. to 1, that would
+# mean that if someone does uf.find(a), they are really querying for
+# uf.find((`a`, 1)). That would return (`a`, 1) because (`a`, 1) is in a new set
+# of its own and hence is the canonical entry of that set.
+#
+# After all of this, we expect the union-find to now hold two sets:
+# {(`a`, 1)} and {(`a`, 0), (`b`, 0)}
+#
+# This situation can be problematic if someone does uf.find(b), i.e.,
+# uf.find((`b`, 0)). Since the canonical entry of the original set continues to
+# point to an older version of `a` i.e. (`a`, 0), there is no tensor for
+# find to retrn, and we raise an error in this case.
+
+# Storing metadata on canonical entries
+# -------------------------------------
+# We associate each set with exactly one metadata dict object. We maintain
+# this invariant by storing a map from canonical tenosors to metadata dict
+# objects. Extra logic is done on-merge in order to maintain this mapping.
+#
+# When two sets A, B are merged, one of the two canonical entries is chosen to
+# be the canonical entry of the union. The other is no longer the canonical
+# entry of any set. To maintain the canonical entry -> metadata dict mapping, we
+# merge A and B's metadata dict objects and store the merged dict on the new
+# canonical entry. The metadata of the no-longer-canonical entry is invalidated.
+#
+# In the case where the two dicts have different entries for
+# the same key, we have a choice of which metadata to favor. In this class, we
+# somewhat arbitrarily chose to favor the metadata of the NON-canonical tensor!
+# This might be the opposite of what one would expect, but that shouldn't matter
+# because which tensor between a and b union-find chooses to be canonical is an
+# implementation detail anyway. The the user should careful to make sure that no
+# such conflict can exist. We could raise an error in this case.
+
+
+class DefaultWeakTensorKeyDictionary(WeakTensorKeyDictionary):
+    def __init__(self, default_cls):
+        super().__init__()
+        self._default_cls = default_cls
+
+    def __getitem__(self, key):
+        if not super().__contains__(key):
+            super().__setitem__(key, self._default_cls())
+        return super().__getitem__(key)
+
+
+class TensorIntMap:
```

Comment:
I guess I haven't read the rest of the PR stack, but it's weird to me that this can live in Python. It doesn't feel like you have enough information in C++.